clang/test/CheckedCRewriter/refarrsubscriptstruct.c

```diff
@@ -0,0 +1,11 @@
+// RUN: cconv-standalone %s -- | FileCheck -match-full-lines %s
+// RUN: cconv-standalone -output-postfix=checked %s 
+// RUN: %clang -c %S/refarrsubscriptstruct.checked.c
+// RUN: rm %S/refarrsubscriptstruct.checked.c
+
+struct foo { int **b; };
+int **bar(struct foo *p) {
+  return &(p->b[1]);
+} 
+//CHECK: struct foo { int **b; };
+//CHECK: int ** bar(_Ptr<struct foo> p) {
```

Comment:
Can you also add tests for when we take the address of an array that's checked? 

And, what about tests with `-alltypes` enabled, so we can see what happens when arrays are being inferred? Are we getting the right answers then?

Reply:
I carried around the implicit assumption that what we have right now for alltypes will work by default because I tunnel-visioned the non-alltypes variant. For this particular test, when run with `-alltypes`, the output is the following:  
```
struct foo { _Array_ptr<_Ptr<int>> b; };
_Nt_array_ptr<_Ptr<int>> bar(_Ptr<struct foo> p) {
  return &(p->b[1]);
}  
```
The above doesn't compile! This means (to address the other comment you made above), there is no fallthrough right now because the assumption that this problem doesn't exist for `-alltypes` is wrong. Working through some examples now, and when I fix them, Ill update the PR with these additional tests.

Reply:
`-alltypes` stuff may not compile, due to missing lengths. You have to ask yourself whether you could manually add lengths, and then it would compile. If the answer is 'yes,' then it's OK. If not, we have a problem ...

Reply:
I think this test result might actually be OK. The reason we have `_Nt_arr_ptr` here might be due to the preference for a *least* solution for return types, which may not always be right. See https://github.com/plum-umd/checkedc-clang/issues/97

Reply:
But even so, I'd prefer there to be an `else` case for the code you added to specifically do the right thing for when we are looking to infer arrays.

Reply:
I'm a little confused about the "right" thing since it seems like we're willing to give `-alltypes` a bit of leeway in some areas. What might the right thing to add to an else clause be?

Reply:
If we're going off some of the current fptrarr tests that were failing before adding this "if" case, it seems to me the right thing is a NoOp, but I feel like I'm misunderstanding your meaning here.

Reply:
The principle I gave above: If it's possible to add the length and the pointer types are correct, then the test is correct. If no such length would allow it to typecheck, then it should not be deemed correct.

Reply:
It's not a "no op" -- *something* is getting returned. What is it? Why is it the right thing?

Reply:
There's no way to react to messages here, but what you're saying makes sense to me. Investigating now!